### DISCUSSION TITLE
Detour a UWP app

### DISCUSSION THREAD

**Discussion by lostmsu** [microsoft/Detours#279] [Q&A]
I am using Detour's `CreateProcessWithDlls` in order to inject a DXGI hook to capture video.

That works great in desktop games, but fails for Minecraft Bedrock Edition, which is an UWP app. Here's what the experience was in steps:

1. I added RX permissions to my hook DLL, so that UWP apps could read and load it (without that the app would not launch in step 2).
2. I called `CreateProcessWithDlls` on Minecraft.Windows.exe
3. Minecraft launched, but my hook was not in the process, and did not work.
4. As a sanity check, I added a Sleep(30000) call to my `DllMain`. What happened is the process did not show up in the Task Manager or in Visual Studio attach dialog until those 30 seconds passed, and when it finally did the DLL was not loaded.

Any suggestions?

**Comment by sylveon** [comment]
Detours [generally does not support UWP](https://github.com/microsoft/Detours/wiki/FAQ#why-cant-my-windows-store-app-for-windows-10-include-detours).

Additionally, launching an exe directly for an UWP app does not work. I suspect that the process you hooked simply relaunched the app properly and terminated itself.

However, if you're simply attempting to capture video, you do not need to hook. All UWP games are borderless fullscreen, so DXGI desktop duplication or Windows.Graphics.Capture should work.

**Reply by lostmsu** [reply]
@sylveon I believe that FAQ entry is only relevant for UWP apps, that include Detours. My app is a desktop app, that wants to inject into a UWP process, so store restrictions do not apply to it. I've seen posts saying that injecting into UWP is generally possible. Question is can it be done with Detours, and what are the steps. You are probably right about relaunching, so it is possible injecting after the launch would work. The `CreateProcessWithDlls` is just cleaner. Perhaps Detours could have an alternative for UWP apps.

The capture and duplication are out of option, as I need this to work when the screen is locked, and they don't.

**Reply by lostmsu** [reply]
Relevant: https://m417z.com/Implementing-Global-Injection-and-Hooking-in-Windows/ Mentions UWP.

**Reply by sylveon** [reply]
It also means that Detours generally isn't tested with UWP, so might not account for any runtime restrictions that are applied to UWP.

**Reply by lostmsu** [reply]
If somebody needs this, I was able to use detours on an UWP app using a combination of [`IApplicationActivationManager`](https://learn.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-iapplicationactivationmanager-activateapplication) and [`IPackageDebugSettings`](https://learn.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ipackagedebugsettings-enabledebugging).

Here's the commit that adds it to my wrapper: https://github.com/losttech/NDetours/commit/bf3601bcaa00a4a6e34edb2f660340d13274aead
